OSPF overhaul with new features, fixes

arifogel opened the pull request:
- Fix broken OSPF inter-area routes
- Add support for OSPF max-metric for:
  - router-lsa
  - stub networks
  - external networks
  - summary networks
- Add support for OSPF point-to-point adjacencies
- Add Cisco support for all of above
- Lots of refactoring

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/batfish/batfish/685)
<!-- Reviewable:end -->

dhalperi commented on projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/Cisco_interface.g4:
```diff
@@ -204,6 +204,11 @@ if_ip_ospf_hello_interval
 
+if_ip_ospf_network
+:
+   IP OSPF NETWORK POINT_TO_POINT NEWLINE
```
is this the only valid line? Otherwise, looks like a regression in the set of things we can handle / ignore.

Or is that also unneeded because the unrecognized line code will handle this anyway (in which case, should we just get rid of the `_null` rules)?

arifogel replied:
`_null` rules are needed so lines don't get thrown out. Not every `_null` rule is an entire line. However those that are could potentially be removed or omitted in future.

Only handling `point-to-point` here is fine since we didn't do anything with other lines anyway.

dhalperi commented on projects/batfish/src/main/java/org/batfish/representation/cisco/OspfProcess.java:
```diff
@@ -21,9 +21,15 @@
 
   private static final OspfMetricType DEFAULT_DEFAULT_INFORMATION_METRIC_TYPE = OspfMetricType.E2;
 
+  public static final long DEFAULT_MAX_METRIC_EXTERNAL_LSA = 16711680L;
```
Would this be better / more clear as `0xFF0000L`?

dhalperi commented on projects/batfish/src/test/java/org/batfish/grammar/cisco/CiscoGrammarTest.java:
```diff
@@ -196,6 +198,77 @@ public void testBgpRemovePrivateAs() throws IOException {
+    long expectedMaxMetricStub = org.batfish.representation.cisco.OspfProcess.MAX_METRIC_ROUTER_LSA;
+    long expectedMaxMetricExternal =
+        org.batfish.representation.cisco.OspfProcess.DEFAULT_MAX_METRIC_EXTERNAL_LSA;
+    long expectedMaxMetricSummary =
```
I wouldn't typically expect you to introduce local constants that exactly match public constants, but ok.

arifogel replied:
It was just to decrease verbosity. Could have used static imports but meh.

dhalperi commented:
Reviewed 71 of 71 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/685)*
<!-- Sent from Reviewable.io -->

dhalperi commented:
*[a discussion](https://reviewable.io:443/reviews/batfish/batfish/685#-KzzTdagHdzGsgB_3rI6:-KzzTdagHdzGsgB_3rI7:b-e5hf7g) (no related file):*
Just leaving a note that reviewable should be considered unresolved based on github comments from commit-by-commit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/685)*
<!-- Sent from Reviewable.io -->

arifogel commented on projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/OspfExternalRoute.java:
```diff
@@ -137,6 +163,10 @@ public long getCostToAdvertiser() {
     return _costToAdvertiser;
   }
 
+  public long getLsaMetric() {
```
Added.

arifogel commented on projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/OspfInternalRoute.java:
```diff
@@ -0,0 +1,28 @@
+    final int prime = 31;
+    int result = 1;
+    result = prime * result + _admin;
+    result = prime * result + (int) (_area ^ (_area >>> 32));
```
Using Objects.hash

arifogel commented:
Review status: 68 of 71 files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/batfish/batfish/685#-KzzTdagHdzGsgB_3rI6:-Kzz_nSWI7Lka-EQKSYj:brywedg) (no related file):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

Just leaving a note that reviewable should be considered unresolved based on github comments from commit-by-commit.
</blockquote></details>

I addressed your comments on github. Please re-review.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/685)*
<!-- Sent from Reviewable.io -->

dhalperi commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/685#-:-KzzdifKeRq3zGWN40Om:bnfp4nl)*
<!-- Sent from Reviewable.io -->
